### PR TITLE
Add cluster setting to dynamically disable filter rewrite optimization

### PR DESCRIPTION
**Pull request by bowenlan-amzn**
Short term solution for pmc workload regression #13171

### Description

Reduce the deciding threshold for rewrite filters from 1024 to 24. Meaning if the date histogram aggregation include more than 24 buckets (e.g. hourly aggregation of 1 day), we won't use the optimization

After this change, we will probably see regression for `date_histogram_hourly_agg` of big5 workload. That will be handled after the long term solution merged in.

### Related Issues
#13171 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- ~[ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java [hunk 1]
```diff
@@ -63,7 +63,7 @@ private FastFilterRewriteHelper() {}
 
     private static final Logger logger = LogManager.getLogger(FastFilterRewriteHelper.class);
 
-    private static final int MAX_NUM_FILTER_BUCKETS = 1024;
+    private static final int MAX_NUM_FILTER_BUCKETS = 24;
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java:66]: I am wondering if we should add setting for number of buckets as well? Might not be widely used, but will be good to not needing code/jar change for testing the right number?

File: server/src/main/java/org/opensearch/search/SearchService.java [hunk 2]
```diff
@@ -272,6 +272,13 @@ public class SearchService extends AbstractLifecycleComponent implements IndexEv
         Property.NodeScope
     );
 
+    public static final Setting<Boolean> FILTER_REWRITE_SETTING = Setting.boolSetting(
+        "search.filter_rewrite.enabled",
+        true,
+        Property.Dynamic,
+        Property.NodeScope
+    );
+
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/search/SearchService.java:281]: I am wondering if there is better name for this setting? Or can we have the number of buckets into this setting to use single setting? Setting to 0 disables the use of rewrite optimization altogether

Comment (bowenlan-amzn) [reply server/src/main/java/org/opensearch/search/SearchService.java]: Discussed offline and updated to this approach. This setting now is an integer, 0 means disable, >0 values will be treated as the max filters threshold to do the optimization.

File: CHANGELOG.md [hunk 3]
```diff
@@ -21,6 +21,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - [Tiered Caching] Add dimension-based stats to ICache implementations. ([#12531](https://github.com/opensearch-project/OpenSearch/pull/12531))
 - Add changes for overriding remote store and replication settings during snapshot restore. ([#11868](https://github.com/opensearch-project/OpenSearch/pull/11868))
 - Add an individual setting of rate limiter for segment replication ([#12959](https://github.com/opensearch-project/OpenSearch/pull/12959))
+- Add cluster setting to dynamically disable filter rewrite optimization. ([#13179](https://github.com/opensearch-project/OpenSearch/pull/13179))
```
Comment (jainankitk) [inline CHANGELOG.md:24]: Minor change to reflect the boolean -> integer change.

```
Add cluster setting to dynamically configure the buckets for filter rewrite optimization.
```

**Review by jainankitk — APPROVED** [review]
LGTM! Also, need to create documentation issue and PR for this change.

**Comment by bowenlan-amzn** [comment]
https://build.ci.opensearch.org/job/gradle-check/37020/

```
org.opensearch.repositories.gcs.GoogleCloudStorageBlobStoreRepositoryTests.testDeleteSingleItem
org.opensearch.repositories.gcs.GoogleCloudStorageBlobStoreRepositoryTests.testDeleteBlobs
org.opensearch.repositories.gcs.GoogleCloudStorageBlobStoreRepositoryTests.testWriteReadLarge
org.opensearch.repositories.gcs.GoogleCloudStorageBlobStoreRepositoryTests.testRequestStats
org.opensearch.repositories.gcs.GoogleCloudStorageBlobStoreRepositoryTests.testIndicesDeletedFromRepository
org.opensearch.repositories.gcs.GoogleCloudStorageBlobStoreRepositoryTests.testMultipleSnapshotAndRollback
org.opensearch.repositories.gcs.GoogleCloudStorageBlobStoreRepositoryTests.testWriteRead
org.opensearch.repositories.gcs.GoogleCloudStorageBlobStoreRepositoryTests.testSnapshotWithLargeSegmentFiles
org.opensearch.repositories.gcs.GoogleCloudStorageBlobStoreRepositoryTests.testList
org.opensearch.repositories.gcs.GoogleCloudStorageBlobStoreRepositoryTests.testContainerCreationAndDeletion
org.opensearch.repositories.gcs.GoogleCloudStorageBlobStoreRepositoryTests.testReadRange
org.opensearch.repositories.gcs.GoogleCloudStorageBlobStoreRepositoryTests.testSnapshotAndRestore
org.opensearch.repositories.gcs.GoogleCloudStorageBlobStoreRepositoryTests.testReadNonExistingPath
```

probably tracked in #11493

**Comment by bowenlan-amzn** [comment]
> LGTM! Also, need to create documentation issue and PR for this change.

Created a documentation PR:
https://github.com/opensearch-project/documentation-website/issues/6957

File: server/src/main/java/org/opensearch/common/settings/ClusterSettings.java [hunk 4]
```diff
@@ -730,7 +730,9 @@ public void apply(Settings value, Settings current, Settings previous) {
 
                 RemoteStoreSettings.CLUSTER_REMOTE_INDEX_SEGMENT_METADATA_RETENTION_MAX_COUNT_SETTING,
                 RemoteStoreSettings.CLUSTER_REMOTE_TRANSLOG_BUFFER_INTERVAL_SETTING,
-                RemoteStoreSettings.CLUSTER_REMOTE_TRANSLOG_TRANSFER_TIMEOUT_SETTING
+                RemoteStoreSettings.CLUSTER_REMOTE_TRANSLOG_TRANSFER_TIMEOUT_SETTING,
+
+                SearchService.MAX_AGGREGATION_REWRITE_FILTERS
```
Comment (msfroh) [inline server/src/main/java/org/opensearch/common/settings/ClusterSettings.java:735]: Can you move this up to around line 522-527 (i.e. with the other `SearchService` settings)?

File: server/src/main/java/org/opensearch/search/SearchService.java [hunk 5]
```diff
@@ -272,6 +272,15 @@ public class SearchService extends AbstractLifecycleComponent implements IndexEv
         Property.NodeScope
     );
 
+    // value 0 means rewrite filters optimization in aggregations will be disabled
+    public static final Setting<Integer> MAX_AGGREGATION_REWRITE_FILTERS = Setting.intSetting(
+        "search.max_aggregation_rewrite_filters",
+        24,
```
Comment (msfroh) [inline server/src/main/java/org/opensearch/search/SearchService.java:278]: Why 24?

If we do a daily aggregation for 1 month, would we want the filter rewrite to kick in? If we do an hourly aggregation for 2 days, should we use filter queries or doc values?

Comment (jainankitk) [reply server/src/main/java/org/opensearch/search/SearchService.java:278]: @bowenlan-amzn - Can we change this to 72 as discussed offline? The value is wide enough to account for some of the cases @msfroh mentioned, but conservative enough to not impact for larger bucket count

File: server/src/main/java/org/opensearch/common/settings/ClusterSettings.java (hunk 4 above)
Comment (jainankitk) [reply server/src/main/java/org/opensearch/common/settings/ClusterSettings.java:735]: My bad, did look odd to me. Thanks for pointing out @msfroh!

File: server/src/main/java/org/opensearch/search/SearchService.java (hunk 5 above)
Comment (bowenlan-amzn) [reply server/src/main/java/org/opensearch/search/SearchService.java:278]: 24 is from the regression of pmc workload, where the optimized method on 24 buckets still perform worse but 12 buckets perform better.
This setting here is easily changeable for open source user and we will have it documented.

File: server/src/main/java/org/opensearch/common/settings/ClusterSettings.java (hunk 4 above)
Comment (bowenlan-amzn) [reply server/src/main/java/org/opensearch/common/settings/ClusterSettings.java]: Updated

File: server/src/main/java/org/opensearch/search/SearchService.java (hunk 5 above)
Comment (bowenlan-amzn) [reply server/src/main/java/org/opensearch/search/SearchService.java]: Updated to 72.

**Comment by bowenlan-amzn** [comment]
[**@github-actions**](https://github.com/apps/github-actions) commented on [Apr 15, 2024, 3:14 PM PDT](https://github.com/opensearch-project/OpenSearch/pull/13179#issuecomment-2057902496 "2024-04-15T22:14:49Z - Replied by Github Reply Comments"):
> :x: Gradle check result for 066739650c2ed1a0ebc35b2204e9c937e730bc9f: [FAILURE](https://build.ci.opensearch.org/job/gradle-check/37043/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?
> 
> _Originally posted by @github-actions[bot] in https://github.com/opensearch-project/OpenSearch/issues/13179#issuecomment-2057902496_

```
org.opensearch.cluster.coordination.AwarenessAttributeDecommissionIT.testConcurrentDecommissionAction
org.opensearch.cluster.coordination.AwarenessAttributeDecommissionIT.testConcurrentDecommissionAction
org.opensearch.cluster.coordination.AwarenessAttributeDecommissionIT.testConcurrentDecommissionAction
org.opensearch.cluster.coordination.AwarenessAttributeDecommissionIT.testConcurrentDecommissionAction
```

Flaky tests tracked in #12197

**Review by msfroh — APPROVED** [review]

**Comment by bowenlan-amzn** [comment]
Manually backport #13217

**Pull request merged**